src/number.rs

```diff
@@ -1,47 +1,327 @@
-use num_bigint::{BigInt, Sign};
-use std::sync::Mutex;
+use std::{fmt, ops::AddAssign};
+
+use ark_ff::{
+    fields::{Field, Fp64, MontBackend, MontConfig},
+    BigInteger, PrimeField, Zero,
+};
+
+#[derive(MontConfig)]
+#[modulus = "18446744069414584321"]
+#[generator = "3"]
+pub struct GoldilocksBaseFieldConfig;
+pub type GoldilocksBaseField = Fp64<MontBackend<GoldilocksBaseFieldConfig, 1>>;
 
 /// The abstract type of numbers to be computed with.
-/// They have arbitrary precision, but need to be converted
-/// to a finite field element once we generate the column values.
-pub type AbstractNumberType = num_bigint::BigInt;
+#[derive(Clone, Copy, PartialEq, Eq, Debug, Default, PartialOrd, Ord, Hash)]
+pub struct AbstractNumberType {
+    value: GoldilocksBaseField,
+}
 /// The type of polynomial degrees and indices into columns.
 pub type DegreeType = u64;
 
-pub fn abstract_to_degree(input: &AbstractNumberType) -> DegreeType {
-    match input.to_biguint().unwrap().to_u64_digits()[..] {
-        [] => 0,
-        [d] => d,
-        _ => panic!(),
+impl AbstractNumberType {
+    pub fn to_degree(&self) -> DegreeType {
+        let value_big = self.value.into_bigint();
+        assert_eq!(value_big.0.len(), 1);
+        value_big.0[0]
+    }
+
+    pub fn modulus() -> DegreeType {
+        GoldilocksBaseField::MODULUS.0[0]
+    }
+
+    pub fn zero() -> Self {
+        Self::from(0)
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.value.is_zero()
+    }
+
+    pub fn pow(self, exponent: u64) -> Self {
+        Self {
+            value: self.value.pow([exponent]),
+        }
+    }
+
+    pub fn integer_div(self, other: Self) -> Self {
+        (self.to_degree() / other.to_degree()).into()
+    }
+
+    pub fn to_bytes_be(&self) -> Vec<u8> {
+        self.value.into_bigint().to_bytes_be()
+    }
+}
+
+impl<V: Into<GoldilocksBaseField>> From<V> for AbstractNumberType {
+    fn from(value: V) -> Self {
+        Self {
+            value: value.into(),
+        }
+    }
+}
+
+// Add
+
+impl std::ops::Add for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        AbstractNumberType {
+            value: self.value + rhs.value,
+        }
+    }
+}
+
+impl<'a> std::ops::Add<&'a AbstractNumberType> for AbstractNumberType {
+    type Output = AbstractNumberType;
+
+    fn add(self, rhs: &'a AbstractNumberType) -> Self::Output {
+        AbstractNumberType {
+            value: self.value + rhs.value,
+        }
+    }
+}
+
+impl<'a> std::ops::Add for &'a AbstractNumberType {
```

Comment:
Why do we need to re-implement all these?

Reply:
Because we're wrapping a type which implements them